File: selfstudy/js/books_avg.js
/*
task 54:
Books - Avg
You need to compute the average of book page counts
*/

function average_books(pages) {
  return pages.length ? pages.reduce((a, b) => a + b, 0) / pages.length : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: average_books([10, 20, 30]) },
  { input: [50, 60, 70], expected: average_books([50, 60, 70]) },
  { input: [1, 2, 3], expected: average_books([1, 2, 3]) },
  { input: [100, 200], expected: average_books([100, 200]) },
  { input: [], expected: average_books([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('average_books', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(average_books(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, average_books(t.input) === t.expected || JSON.stringify(average_books(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// done

File: selfstudy/golang/books_avg.go
/*
task 254:
Books - Avg
You need to compute the average of book page counts
*/

package main

import (
	"fmt"
	"reflect"
)

func AverageBooks(pages []int) float64 {
	if len(pages) == 0 { return 0 }
	sum := 0
	for _, v := range pages {
		sum += v
	}
	return float64(sum)/float64(len(pages))
}

func main() {
	tests := []struct {
		input []int
		expected float64
	}{
		{[]int{10, 20, 30}, AverageBooks([]int{10, 20, 30})},
		{[]int{50, 60, 70}, AverageBooks([]int{50, 60, 70})},
		{[]int{1, 2, 3}, AverageBooks([]int{1, 2, 3})},
		{[]int{100, 200}, AverageBooks([]int{100, 200})},
		{[]int{}, AverageBooks([]int{})},
	}
	
	for idx, t := range tests {
		res := AverageBooks(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// done
